docs/standard/serialization/system-text-json-converters-how-to.md

```diff
@@ -0,0 +1,745 @@
+                        v = _valueConverter.Read(ref reader, _valueType, options);
+                    }
+                    else
+                    {
+                        v = JsonSerializer.Deserialize<TValue>(ref reader, options);
+                    }
+
+                    // Add to dictionary.
+                    value.Add(key, v);
+                }
+
+                throw new JsonException();
+            }
+
+            public override void Write(
+                Utf8JsonWriter writer, 
+                Dictionary<TKey, TValue> value, 
+                JsonSerializerOptions options)
+            {
+                writer.WriteStartObject();
+
+                foreach (KeyValuePair<TKey, TValue> kvp in value)
+                {
+                    writer.WritePropertyName(kvp.Key.ToString());
+
+                    if (_valueConverter != null)
+                    {
+                        _valueConverter.Write(writer, kvp.Value, options);
+                    }
+                    else
+                    {
+                        JsonSerializer.Serialize(writer, kvp.Value, options);
+                    }
+                }
+
+                writer.WriteEndObject();
+            }
+        }
+    }
+}
+```
+
+The preceding code follows the factory pattern because the first generic type parameter is `Enum` and the second is open. The `CanConvert` method returns `true` only for a `Dictionary` with two generic parameters, the first of which is an `Enum` type. The inner converter gets an existing converter to handle whichever type is provided at runtime for `TValue`. 
+
+The following code registers the converter:
+
+```csharp
+var serializeOptions = new JsonSerializerOptions();
+serializeOptions.Converters.Add(new ConverterDictionaryTKeyEnumTValue());
+```
+
+The converter can serialize and deserialize the `TemperatureRanges` property of the following class that uses the following `Enum`:
+
+```csharp
+public class WeatherForecastWithEnumDictionary
+{
+    public DateTimeOffset Date { get; set; }
+    public int TemperatureC { get; set; }
+    public string Summary { get; set; }
+    public Dictionary<SummaryWordsEnum, int> TemperatureRanges { get; set; }
+}
+
+public enum SummaryWordsEnum
+{
+    Cold, Hot
+}
+```
+
+The JSON output from serialization looks like the following example:
+
+```json
+{
+
+  "Date": "2019-08-01T00:00:00-07:00",
+  "TemperatureC": 25,
+  "Summary": "Hot",
+  "TemperatureRanges": {
+    "Cold": 20,
+    "Hot": 40
+  }
+}
+```
+
+The [unit tests folder](https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/tests/Serialization/) in the `System.Text.Json.Serialization` namespace has more examples of custom converters that handle non-string-key dictionaries.
+
+### Support polymorphic deserialization
+
+[Polymorphic serialization](system-text-json-how-to.md#serialize-properties-of-derived-classes) is supported, but a custom converter is required for deserialization.
```

Comment:
Polymorphic serialization is not truly supported by default -- it is only supported if the property's type is System.Object. This is for secure-by-default and because we'd have no way to deserialize \ round-trip the type since the type information is not specified in the JSON in some way.

However we want to have an opt-in option for 5.0 to enable polymorphic serialization for any type but also want to ensure we have a pattern or support for polymorphic deserialization (such as including a type discriminator as metadata in the JSON in a secure way).